intermediate_source/speech_command_recognition_with_torchaudio.py

```diff
@@ -0,0 +1,523 @@
+        x = self.pool3(x)
+        x = self.conv4(x)
+        x = F.relu(self.bn4(x))
+        x = self.pool4(x)
+        x = F.avg_pool1d(x, x.shape[-1])
+        x = x.permute(0, 2, 1)
+        x = self.fc1(x)
+        return F.log_softmax(x, dim=2)
+
+
+model = M5(n_input=transformed.shape[0], n_output=len(labels))
+model.to(device)
+print(model)
+
+
+def count_parameters(model):
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+
+n = count_parameters(model)
+print("Number of parameters: %s" % n)
+
+
+######################################################################
+# We will use the same optimization technique used in the paper, an Adam
+# optimizer with weight decay set to 0.0001. At first, we will train with
+# a learning rate of 0.01, but we will use a ``scheduler`` to decrease it
+# to 0.001 during training after 20 epochs.
+#
+
+optimizer = optim.Adam(model.parameters(), lr=0.01, weight_decay=0.0001)
+scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=20, gamma=0.1)  # reduce the learning after 20 epochs by a factor of 10
+
+
+######################################################################
+# Training and Testing the Network
+# --------------------------------
+#
+# Now let’s define a training function that will feed our training data
+# into the model and perform the backward pass and optimization steps. For
+# training, the loss we will use is the negative log-likelihood. The
+# network will then be tested after each epoch to see how the accuracy
+# varies during the training.
+#
+
+
+def train(model, epoch, log_interval):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+
+        data = data.to(device)
+        target = target.to(device)
+
+        # apply transform and model on whole batch directly on device
+        data = transform(data)
+        output = model(data)
+
+        # negative log-likelihood for a tensor of size (batch x 1 x n_output)
+        loss = F.nll_loss(output.squeeze(), target)
+
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        # print training stats
+        if batch_idx % log_interval == 0:
+            print(f"Train Epoch: {epoch} [{batch_idx * len(data)}/{len(train_loader.dataset)} ({100. * batch_idx / len(train_loader):.0f}%)]\tLoss: {loss:.6f}")
+
+        if "pbar" in globals() and "pbar_update" in globals():
+            pbar.update(pbar_update)
+
+
+######################################################################
+# Now that we have a training function, we need to make one for testing
+# the networks accuracy. We will set the model to ``eval()`` mode and then
+# run inference on the test dataset. Calling ``eval()`` sets the training
+# variable in all modules in the network to false. Certain layers like
+# batch normalization and dropout layers behave differently during
+# training so this step is crucial for getting correct results.
+#
+
+
+def argmax(tensor):
```

Comment:
What is the advantage of this custom `argmax` over `torch.argmax`?

Reply:
Good catch, removed function in favor of torch.argmax.